File: backend/src/main/java/com/tunisair/meeting/service/DocumentService.java
package com.tunisair.meeting.service;

import com.tunisair.meeting.models.Document;
import com.tunisair.meeting.models.DocumentVersion;
import com.tunisair.meeting.models.Meeting;
import com.tunisair.meeting.models.User;
import com.tunisair.meeting.repository.DocumentRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.multipart.MultipartFile;

import java.util.List;
import java.util.Optional;

@Service
public class DocumentService {
    @Autowired
    private DocumentRepository documentRepository;

    @Transactional
    public Document uploadDocument(MultipartFile file, Meeting meeting, User uploadedBy, String description) {
        // Implementation of uploadDocument method
        return null; // Placeholder return, actual implementation needed
    }

    @Transactional(readOnly = true)
    public Optional<Document> getDocumentById(Long id) {
        return documentRepository.findById(id);
    }

    @Transactional(readOnly = true)
    public List<Document> getAllDocuments() {
        return documentRepository.findAll();
    }

    @Transactional
    public Document updateDocument(Document document) {
        return documentRepository.save(document);
    }

    @Transactional
    public void deleteDocument(Long id) {
        documentRepository.deleteById(id);
    }

    @Transactional(readOnly = true)
    public List<Document> getDocumentsByMeeting(Meeting meeting) {
        return documentRepository.findByMeeting(meeting);
    }

    @Transactional(readOnly = true)
    public List<Document> getDocumentsByUploader(User uploader) {
        return documentRepository.findByUploader(uploader);
    }

    @Transactional(readOnly = true)
    public List<Document> getDocumentsByMeetingAndType(Meeting meeting, String fileType) {
        return documentRepository.findByMeetingAndFileType(meeting, fileType);
    }

    @Transactional(readOnly = true)
    public List<Document> getLatestDocumentsByMeeting(Meeting meeting) {
        return documentRepository.findByMeetingOrderByUploadDateDesc(meeting);
    }

    @Transactional(readOnly = true)
    public List<Document> searchDocuments(String keyword) {
        // Implementation of searchDocuments method
        return null; // Placeholder return, actual implementation needed
    }

    @Transactional
    public Document updateDocumentVersion(Long documentId, MultipartFile newFile, String description) {
        // Implementation of updateDocumentVersion method
        return null; // Placeholder return, actual implementation needed
    }

    @Transactional
    public byte[] downloadDocument(Long documentId) {
        // Implementation of downloadDocument method
        return null; // Placeholder return, actual implementation needed
    }

    // Version control methods
    @Transactional(readOnly = true)
    public List<DocumentVersion> getDocumentVersions(Long documentId) {
        // Implementation of getDocumentVersions method
        return null; // Placeholder return, actual implementation needed
    }

    @Transactional(readOnly = true)
    public DocumentVersion getCurrentVersion(Long documentId) {
        // Implementation of getCurrentVersion method
        return null; // Placeholder return, actual implementation needed
    }

    @Transactional(readOnly = true)
    public DocumentVersion getVersionByNumber(Long documentId, String versionNumber) {
        // Implementation of getVersionByNumber method
        return null; // Placeholder return, actual implementation needed
    }

    @Transactional
    public DocumentVersion revertToVersion(Long documentId, String versionNumber) {
        // Implementation of revertToVersion method
        return null; // Placeholder return, actual implementation needed
    }

    @Transactional
    public void deleteVersion(Long documentId, String versionNumber) {
        // Implementation of deleteVersion method
    }
} 

File: frontend/src/app/document/services/document.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';
import { Document, DocumentVersion } from '../models/document.model';

@Injectable({
  providedIn: 'root'
})
export class DocumentService {
  private apiUrl = `${environment.apiUrl}/documents`;

  constructor(private http: HttpClient) { }

  uploadDocument(file: File, meetingId: number, description: string): Observable<Document> {
    const formData = new FormData();
    formData.append('file', file);
    formData.append('meetingId', meetingId.toString());
    formData.append('description', description);
    return this.http.post<Document>(`${this.apiUrl}/upload`, formData);
  }

  updateDocumentVersion(documentId: number, file: File, description: string): Observable<Document> {
    const formData = new FormData();
    formData.append('file', file);
    formData.append('description', description);
    return this.http.put<Document>(`${this.apiUrl}/${documentId}/version`, formData);
  }

  deleteDocument(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }

  getDocumentById(id: number): Observable<Document> {
    return this.http.get<Document>(`${this.apiUrl}/${id}`);
  }

  getDocumentsByMeeting(meetingId: number): Observable<Document[]> {
    return this.http.get<Document[]>(`${this.apiUrl}/meeting/${meetingId}`);
  }

  getDocumentsByUploader(userId: number): Observable<Document[]> {
    return this.http.get<Document[]>(`${this.apiUrl}/uploader/${userId}`);
  }

  searchDocuments(keyword: string): Observable<Document[]> {
    return this.http.get<Document[]>(`${this.apiUrl}/search`, {
      params: { keyword }
    });
  }

  downloadDocument(documentId: number): Observable<Blob> {
    return this.http.get(`${this.apiUrl}/${documentId}/download`, {
      responseType: 'blob'
    });
  }

  getDocumentVersions(documentId: number): Observable<DocumentVersion[]> {
    return this.http.get<DocumentVersion[]>(`${this.apiUrl}/${documentId}/versions`);
  }

  getCurrentVersion(documentId: number): Observable<DocumentVersion> {
    return this.http.get<DocumentVersion>(`${this.apiUrl}/${documentId}/versions/current`);
  }

  getVersionByNumber(documentId: number, versionNumber: string): Observable<DocumentVersion> {
    return this.http.get<DocumentVersion>(`${this.apiUrl}/${documentId}/versions/${versionNumber}`);
  }

  revertToVersion(documentId: number, versionNumber: string): Observable<Document> {
    return this.http.put<Document>(`${this.apiUrl}/${documentId}/versions/${versionNumber}/revert`, {});
  }

  deleteVersion(documentId: number, versionNumber: string): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${documentId}/versions/${versionNumber}`);
  }
} 